### PR TITLE
categorical distribution

### PR DESCRIPTION
**Pull request by mayalenE**
Hi @lockwo, here is a new PR to add the categorical distribution :slightly_smiling_face: 

Two things i'm not entirely happy about:
- I've reproduced the many tests from distrax akin to what you did for the bernouilli distribution but I feel some of these tests are pretty limited with respect to the original ones as at the end we are only comparing the shape whereas it would be nice to check that the computed values are the same than when compared with distrax and/or tfp. Although this would need to install those packages adding an additional dependency, which is why I've left them as this (with TODO comments) for the moment.
- Due to the pyright pre-commit I needed to check that `_logits` and `_probs` are not None in several functions but this is already checked in the constructor, adding useless compute. I see that you did the same in bernouilli so also left it like this for the moment (with TODO comments).

What do you think?

Now I will work on adding the one_hot_categorical distribution,
Best,
Maya

### CODE REVIEW

**Comment by lockwo** [comment]
I agree 100% on the need for more tests (https://github.com/lockwo/distreqx/issues/15), although I am not sure about testing against distrax or TFP (that requires kind of clunky tests and requires people who want to contribute to this package to figure out additional packages). Personally, I lean towards adding tests for known solutions (i.e. test against things we can analytically solve for to verify correctness of implementation)

**Comment by lockwo** [comment]
The if None is annoying I agree, in terms of LoC just doing `assert x is not None` might be shorter (assuming we know it shouldn't be at that point). That's sort of the downside of not materializing both probs and logits in init, which maybe we can just do

File: distreqx/distributions/categorical.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array, PRNGKeyArray
+
+from ..utils.math import mul_exp, multiply_no_nan, normalize
+from ._distribution import AbstractDistribution
+
+
+class Categorical(AbstractDistribution):
+    """Categorical distribution."""
```
Comment (lockwo) [inline distreqx/distributions/categorical.py:14]: Could be good to add some description, i.e. so people know that this distribution describes a single categorical variable and such

File: distreqx/distributions/categorical.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        if self._logits is None:
+            raise ValueError(
+                "_probs and _logits are None!"
+            )  # TODO: useless but needed for pyright
+        return self._logits.shape[-1]
+
+    def sample(self, key: PRNGKeyArray) -> Array:
+        """See `Distribution.sample`."""
+        is_valid = jnp.logical_and(
+            jnp.all(jnp.isfinite(self.probs), axis=-1),
```
Comment (lockwo) [inline distreqx/distributions/categorical.py:82]: Do we need the axis=-1 here? If the event shape is (), then don't these have to be 1D arrays?

File: distreqx/distributions/categorical.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            multiply_no_nan(jnp.cumsum(self.probs, axis=-1), value_one_hot), axis=-1
+        )
+        return jnp.where(should_be_zero, 0.0, jnp.where(should_be_one, 1.0, cdf))
+
+    def log_cdf(self, value: Array) -> Array:
+        """See `Distribution.log_cdf`."""
+        return jnp.log(self.cdf(value))
+
+    def logits_parameter(self) -> Array:
+        """Wrapper for `logits` property, for TFP API compatibility."""
```
Comment (lockwo) [inline distreqx/distributions/categorical.py:156]: We probably don't need this if its just for the TFP API

File: tests/categorical_test.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+class CategoricalTest(TestCase):
+    def setUp(self):
+        self.dist = categorical.Categorical
+        self.key = jax.random.PRNGKey(0)
+
+    @parameterized.expand(
+        [
+            ("1d probs", (4,), True),
+            ("1d logits", (4,), False),
+            ("2d probs", (3, 4), True),
```
Comment (lockwo) [inline tests/categorical_test.py:24]: Related to an early comment, is the 3 here just an implicit batch? Like if we have event size == () and this array doesn't get flattened, then its just 3 separate () sized events?

**Comment by lockwo** [comment]
Also, I see one of the categorical tests is failing. Is something you can replicate locally?

**Comment by mayalenE** [comment]
Hi @lockwo coming back to you wrt my previous comments, did you maybe have a chance to look at them? Would be very helpful to update the PR :pray: 
In particular I'm not sure for the event shape if I keep () or rather put prob.shape[-1] akin to the bernouilli distrib (batch dim)

Thank you again!

**Comment by lockwo** [comment]
> Hi @lockwo coming back to you wrt my previous comments, did you maybe have a chance to look at them? Would be very helpful to update the PR 🙏 In particular I'm not sure for the event shape if I keep () or rather put prob.shape[-1] akin to the bernouilli distrib (batch dim)
> 
> Thank you again!

Which comments should I be looking at wrt? I see this PR which I left comments on and this seems to be the most up to date version

**Comment by lockwo** [comment]
As for "In particular I'm not sure for the event shape if I keep () or rather put prob.shape[-1] akin to the bernouilli distrib (batch dim)" I think both could be valid, basically, we just want the event shape to be the shape of the distribution. Bernoulli is like that because the distribution could be higher dimensional. Like unbatched shape of the distribution. I guess there is an implicit batching for these sort of distributions but we have to consider when people use independent how it should interact. Does that make sense?

**Comment by lockwo** [comment]
To put it another way. Categorical distributions are distributions that describe a random variable that can take K values. However, we can of course generalize this to vectors of random variables and describe this as the distribution. Concretely, this is kind of like a batch, since each element is totally independent. I'm leaning toward unofficially supporting it (since it seems trivial to generalize it at least in the Bernoulli case since every op is element wise), but officially recommending people vmap over initialization (which would product identical results) and vmap over ops as well if they need batches.

**Comment by mayalenE** [comment]
Thanks for the answer @lockwo ! 

> I'm leaning toward unofficially supporting it (since it seems trivial to generalize it at least in the Bernoulli case since every op is element wise), but officially recommending people vmap over initialization (which would product identical results) and vmap over ops as well if they need batches.

Yes that makes sense I will leave the "batch" dim (or call it differently) for the moment then.

That being said, I think it might be confusing for users that for some distribs we can "batch" at initialization and for others we need to vmap. For instance I have a use case with a GMM where I need to define an ensemble of `MultivariateNormalTri` distribs which is feasible with vmap but more cumbersome to write than simply initializing with batched 'loc' and 'scale_tri', so can be confusing if for some distribs we can get around this need. But probably i'm missing the reason why you decided to remove the batch dim originally. 

> Which comments should I be looking at wrt? I see this PR which I left comments on and this seems to be the most up to date version

I'm reffering to reviews that I've made a review on my  PR (or comments to your reviews)

**Comment by lockwo** [comment]
Looks good, I think if you bring it up with main and fix the merge commit it will be good. I also used a variant of your code, so you can see how I brought it up with main here: https://github.com/lockwo/distreqx/blob/Owen/add-mixture/distreqx/distributions/categorical.py
